ip_country/src/bit_queue.rs

```diff
@@ -0,0 +1,413 @@
+                bit_count
+            )
+        }
+        if bit_count > self.len() {
+            return None;
+        }
+        let mut bit_data = 0u64;
+        let mut bit_position = 0usize;
+        let (initial_bit_data, initial_bit_count) = self.take_some_front_bits(bit_count);
+        bit_data |= initial_bit_data << bit_position;
+        bit_position += initial_bit_count;
+        bit_count -= initial_bit_count;
+        let (byte_bit_data, byte_bit_count) = self.take_front_bytes(bit_count);
+        bit_data |= byte_bit_data << bit_position;
+        bit_position += byte_bit_count;
+        bit_count -= byte_bit_count;
+        let (final_front_bit_data, final_front_bit_count) = self.take_some_front_bits(bit_count);
+        if final_front_bit_count > 0 {
+            bit_data |= final_front_bit_data << bit_position;
+            bit_position += final_front_bit_count;
+        }
+        bit_count -= final_front_bit_count;
+        let (final_back_bit_data, final_back_bit_count) = self.take_some_back_bits(bit_count);
+        if final_back_bit_count > 0 {
+            bit_data |= final_back_bit_data << bit_position;
+            bit_position += final_back_bit_count;
+        }
+        bit_count -= final_back_bit_count;
+        if bit_position != original_bit_count {
+            panic!(
+                "Wanted {} bits, but got {} instead",
+                original_bit_count, bit_position
+            );
+        }
+        return Some(bit_data);
+    }
+
+    fn back_full_bit_count(&self) -> usize {
+        8 - self.back_blank_bit_count
+    }
+
+    fn front_full_bit_count(&self) -> usize {
+        8 - self.front_blank_bit_count
+    }
+
+    fn low_order_ones(count: usize) -> u64 {
+        !(u64::MAX << count)
+    }
+
+    #[allow(dead_code)]
+    fn dump_queue(&self) -> String {
+        let queue_str = self
+            .byte_queue
+            .iter()
+            .map(|b| format!("{:08b}", *b))
+            .rev()
+            .collect::<Vec<String>>()
+            .join(" ");
+        format!(
+            "{}->{}->{}",
+            self.back_blank_bit_count, queue_str, self.front_blank_bit_count
+        )
+    }
+
+    fn add_some_back_bits(&mut self, bit_data: u64, bit_count: usize) -> usize {
+        if self.back_blank_bit_count == 0 {
+            self.byte_queue.push_back(0);
+            self.back_blank_bit_count = 8;
+        }
+        let bits_to_add = bit_count.min(self.back_blank_bit_count);
+        let back_full_bit_count = self.back_full_bit_count();
+        let back_ref = self
+            .byte_queue
+            .back_mut()
+            .expect("There should be a back byte");
+        *back_ref |= (bit_data << back_full_bit_count) as u8;
+        self.back_blank_bit_count -= bits_to_add;
+        if self.back_blank_bit_count == 0 {
+            self.byte_queue.push_back(0);
+            self.back_blank_bit_count = 8;
+        }
```

Comment:
These lines repeat as well. 

1) lns. 117 - 120
2) lns. 129 - 132

It should probably be a special method.